File: bubble/bubble.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

// Função Bubble Sort simples
void bubble_sort(int *array, int n) {
    for (int i = 0; i < n-1; i++) {
        for (int j = 0; j < n-i-1; j++) {
            if (array[j] > array[j+1]) {
                int tmp = array[j];
                array[j] = array[j+1];
                array[j+1] = tmp;
            }
        }
    }
}

int main(int argc, char** argv) {
    int rank, size, n = 16; 
    int *data = NULL;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    int local_n = n / size;
    int *local_data = (int*)malloc(local_n * sizeof(int));


    if (rank == 0) {
        data = (int*)malloc(n * sizeof(int));
        printf("Vetor original:\n");
        for (int i = 0; i < n; i++) {
            data[i] = rand() % 100;
            printf("%d ", data[i]);
        }
        printf("\n");
    }


    MPI_Scatter(data, local_n, MPI_INT, local_data, local_n, MPI_INT, 0, MPI_COMM_WORLD);

    bubble_sort(local_data, local_n);

    MPI_Gather(local_data, local_n, MPI_INT, data, local_n, MPI_INT, 0, MPI_COMM_WORLD);

    if (rank == 0) {
        bubble_sort(data, n);

        printf("Vetor ordenado final:\n");
        for (int i = 0; i < n; i++) {
            printf("%d ", data[i]);
        }
        printf("\n");

        free(data);
    }

    free(local_data);
    MPI_Finalize();
    return 0;
}


================
Compile e execute o código Bubble Sort com MPI fornecido.
2 Modifique o vetor para os tamanhos: 32, 64, 128 e 256 elementos.
3 Execute com 2, 4 e 8 processos, e registre o tempo de execução.
4 Faça cada processo imprimir seu vetor local antes e depois da
ordenação.
5 (Opcional) Implemente uma fusão (merge) no processo 0 para
substituir a ordenação final.
6 Responda:
Qual o impacto do número de processos no tempo?
Em quais situações houve ganho real?
Quais limitações você identificou?
Como melhorar a eficiência do algoritmo?

File: bubble/python/bubble.py
from mpi4py import MPI
import numpy as np
import time

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

N = 128  # Altere para 32, 64, 128 ou 256

local_n = N // size
local_data = np.empty(local_n, dtype='i')

if rank == 0:
    data = np.random.randint(0, 100, size=N, dtype='i')
    print("Vetor original:", data)
else:
    data = None

comm.Scatter(data, local_data, root=0)

print(f"[Rank {rank}] Antes da ordenação: {local_data}")

start_time = time.time()

bubble_sort(local_data)

print(f"[Rank {rank}] Depois da ordenação: {local_data}")

gathered_data = None
if rank == 0:
    gathered_data = np.empty(N, dtype='i')

comm.Gather(local_data, gathered_data, root=0)

if rank == 0:
    bubble_sort(gathered_data)
    end_time = time.time()
    print("Vetor ordenado final:", gathered_data)
    print(f"Tempo total: {end_time - start_time:.6f} segundos")

